### ISSUE TITLE
Restrict JPEG to JFIF

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/glTF#1878]
Although IANA (RFC 2046) and W3C agree that `image/jpeg` technically means "JPEG format using JFIF encoding", many software packages can export `.jpg` files using other profiles such as EXIF or even include proprietary Photoshop metadata. This turns out to be a compatibility / portability issue as support for those advanced features varies across platforms.

So the suggestion is to state that only JPEG JFIF is supported in glTF and to call out the following restrictions. Note, that all of them are already implicitly present in the spec because of `image/jpeg` reference and glTF's color space definitions.
- number of components must be 1 (grayscale) or 3 (color);
- 8-bit color depth;
- although JFIF uses BT.601 color primaries, glTF implementations interpret RGB values as using BT.709 color primaries;
- when present, embedded ICC profile in APP2 must be ignored.

/cc @bghgary @donmccurdy @emackey @MarkCallow

**Comment by MarkCallow** [comment]
JFIF also specifies a linear transfer function.

I think we have to support EXIF given that is what most digital cameras produce. We can limit it to those which are sRGB.

**Comment by lexaknyazev** [comment]
Reliably supporting EXIF (i.e. providing normative language and test cases) would be a substantial amount of work given all its options. glTF is not meant to be used with untouched digital camera's output.

JFIF's linear transfer requirement is ignored by everyone. In practice, JPEG implementations treat RGB values recovered from YCbCr as either having transfer function provided via APP2 or as using sRGB.

**Comment by donmccurdy** [comment]
EXIF metadata would still be okay, as long as it's not changing the encoding, correct? Taking a simple screenshot on macOS and converting it from PNG to JPEG does preserve some trivial EXIF metadata:

```
EXIF

Orientation
    Horizontal (normal)
User Comment
    Screenshot
Exif Image Width
    287
Exif Image Height
    264
```

If graphics APIs do not consistently support EXIF encoding, then I don't think the fact that digital cameras might output it should affect our decision.

**Comment by lexaknyazev** [comment]
> EXIF metadata would still be okay, as long as it's not changing the encoding, correct?

Here's what Wikipedia says about it:
> The newer Exchangeable image file format (Exif) is comparable to JFIF, but the two standards are mutually incompatible. This is because both standards specify that their particular application segment (APP0 for JFIF, APP1 for Exif) must immediately follow the SOI marker. In practice, many programs and digital cameras produce files with both application segments included. This will not affect the image decoding for most decoders, but poorly designed JFIF or Exif parsers may not recognize the file properly.

> If graphics APIs do not consistently support EXIF encoding...

JPEG decoding is handled by web-browsers / platform SDKs rather than GPUs.

**Comment by donmccurdy** [comment]
> JPEG decoding is handled by web-browsers / platform SDKs rather than GPU APIs.

No glTF implementation will re-implement JPEG decoding from scratch — so the first question is, do common JPEG decoders (in browsers, and elsewhere) support EXIF? If so — and to be clear, I don't know — is there any benefit to banning a portion of the JPEG spec that is already universally supported?

> ... In practice, many programs and digital cameras produce files with both application segments included. This will not affect the image decoding for most decoders, but poorly designed JFIF or Exif parsers may not recognize the file properly.

So it sounds like JFIF should be required, but the presence of additional EXIF metadata should not be prohibited? To post a quick example:

![jifif_with_exif_metadata](https://user-images.githubusercontent.com/1848368/95400647-f78e3580-08bf-11eb-8749-198480467bfb.jpg)

According to https://exifinfo.org/detail/Y1NQL3bCgYi3HEYqsMHqBg, this image (assuming GitHub doesn't re-compress it) uses JFIF version 1.01, but also contains EXIF metadata unrelated to the encoding.

**Comment by lexaknyazev** [comment]
The file above contains a JFIF tag in a proper place (right after the SOI), so it would pass as a valid JFIF. The presence of EXIF header in APP2 should have no effect on the file being JFIF. At the same time, glTF should give no warranties about EXIF metadata, so such file may be flagged during validation.

There could also be JPEG files (e.g. using CMYK-like color models) that do not have JFIF header at all. They clearly violate IANA / W3C definition of `image/jpeg` and should also be flagged in my opinion.

**Comment by donmccurdy** [comment]
^This all sounds OK to me. 👍

**Comment by MarkCallow** [comment]
@lexaknyazev wrote that JFIF's linear transfer requirement is ignored by everyone. (What concrete evidence is there for that?) If it is not true then presence of an EXIF APP1 chunk says that the data is sRGB encoded, unless the EXIF metadata contains an ICC profile. As has been noted, what is important here is what Browser's and WebGL do when decoding and uploading a JPEG image from a JS object such as an HTMLImageElement. How can we find out?

**Comment by lexaknyazev** [comment]
> What concrete evidence is there for that?

Here's a simple test. The original file posted above has the following header structure (full contents omitted for brevity):
```
*** Marker: SOI (xFFD8) ***
  OFFSET: 0x00000000
 
*** Marker: APP0 (xFFE0) ***
  OFFSET: 0x00000002
  Length     = 16
  Identifier = [JFIF]
 
*** Marker: APP1 (xFFE1) ***
  OFFSET: 0x00000014
  Length          = 106
  Identifier      = [Exif]

*** Marker: APP1 (xFFE1) ***
  OFFSET: 0x00000080
  Length          = 2337
  Identifier      = [http://ns.adobe.com/xap/1.0/]
    XMP = 
 
*** Marker: APP13 (xFFED) ***
  OFFSET: 0x000009A3
  Length          = 56
  Identifier      = [Photoshop 3.0]
 
*** Marker: APP2 (xFFE2) ***
  OFFSET: 0x000009DD
  Length          = 3420
  Identifier      = [ICC_PROFILE]
```
As you can see, it has JFIF in APP0, Exif in APP1, and an ICC profile (regular sRGB) in APP2. XMP and Photoshop payloads could be obviously ignored here.

By using a hex editor, I removed XMP and Photoshop sections first and made three new files by selectively removing other headers while keeping the pixel data intact:

**JFIF-only (removed both Exif and ICC)**
![jfif-only](https://user-images.githubusercontent.com/3479527/95475398-3cd16680-0997-11eb-985d-4cf7feda4079.jpg)

**JFIF + Exif (removed ICC)**
![jfif+exif](https://user-images.githubusercontent.com/3479527/95475729-976ac280-0997-11eb-8fe8-5e58098ff5b9.jpg)

**JFIF + ICC (removed Exif)**
![jfif+icc](https://user-images.githubusercontent.com/3479527/95475789-a782a200-0997-11eb-9bec-abe16651144b.jpg)

**Comment by lexaknyazev** [comment]
My revised proposal for the glTF validation is:
- Validation error when JFIF marker is absent.
- Validation warning when ICC is present (as we do already for PNG).
- Validation info when Exif is present (also useful for flagging potential PII leaks).

> what Browser's and WebGL do when decoding and uploading a JPEG image

The only control knob is the `UNPACK_COLORSPACE_CONVERSION_WEBGL` parameter that can be set with `pixelStorei` call. It has two options: _default_ (means let the browser decide on color conversions) and _none_ (means passing RGBA data intact). glTF spec already asks engines to use the latter option.

**Comment by lexaknyazev** [comment]
I'd suggest downloading the files locally and comparing them using some imaging software rather than relying on perceived differences on a screen (especially in vertical direction). For example, ImageMagick reports zero MSE and Infinity PSNR for all image pairs above.

To directly check browser's decoding, a simple JavaScript program could be written to get 8-bit pixel values from the Image element.

**Comment by MarkCallow** [comment]
I used the macOS ColorSync utility to look at the RGB values being displayed by Firefox. JFIF-only and JFIF + Exif are the same. JFIF + ICC is definitely different. When I open the last in Corel PaintShop Pro it tells me there is a color space mismatch and that it is converting the image to the working space sRGB. It doesn't give any such message for the first two.

**Comment by lexaknyazev** [comment]
>  JFIF-only and JFIF + Exif are the same

This confirms that JFIF's linear transfer requirement is ignored.

> JFIF + ICC is definitely different.

Just by any chance, is your system configured to use Display P3 color space (it may actually be the default value)? If so, that could be the reason why Firefox shows different colors. Please also check in Safari.

---

Anyway, glTF doesn't care about any of this as the transfer function is always enforced by a material's texture slot and all browser's conversions should be disabled.

**Comment by MarkCallow** [comment]
I have my system configured to Adobe RGB. That is not relevant to the behavior of PaintShop Pro finding a need to convert the image to its working space on loading it.

In Safari all 3 images look the same and like the Firefox rendering of JFIF + ICC.

**Comment by lexaknyazev** [comment]
- Firefox likely gets confused by AdobeRGB system color space and does something wrong.
- Upon further investigation, it's clear that the ICC profile payload in the image above stores the sRGB transfer function using a precomputed table with 1024 values instead of a parametric curve. Likely for that reason, PaintShop Pro doesn't try to recognize the curve and suggests conversion.

**Comment by lexaknyazev** [comment]
@MarkCallow 
From the latest Firefox 89 release notes:
> Colors in Firefox on macOS will no longer be saturated on wide gamut displays, untagged images are properly treated as sRGB, and colors in images tagged as sRGB will now match CSS colors.

**Comment by MarkCallow** [comment]
After updating to FF 89, the JFIF+EXIF and JFIF+ICC samples look identical. I have the impression that the JFIF-only sample is ever so slightly lighter than the other 2 samples - the dark areas don't seem quite as dark. However it is so close it could just be my imagination.

**Comment by lexaknyazev** [comment]
Addressed in #1997.
